source/common/tracing/http_tracer_impl.cc

```diff
@@ -249,5 +265,86 @@ SpanPtr HttpTracerImpl::startSpan(const Config& config, Http::HeaderMap& request
   return active_span;
 }
 
+void GeneralCustomTag::apply(Span& span, const CustomTagContext& ctx) const {
+  const absl::string_view& tag_value = value(ctx);
+  if (!tag_value.empty()) {
+    span.setTag(tag(), tag_value);
+  }
+}
+
+EnvironmentCustomTag::EnvironmentCustomTag(
+    const std::string& tag, const envoy::type::tracing::v2::CustomTag::Environment& environment)
+    : GeneralCustomTag(tag), name_(environment.name()),
+      default_value_(environment.default_value()) {
+  const char* env = std::getenv(name_.data());
+  final_value_ = env ? env : default_value_;
+}
+
+RequestHeaderCustomTag::RequestHeaderCustomTag(
+    const std::string& tag, const envoy::type::tracing::v2::CustomTag::Header& request_header)
+    : GeneralCustomTag(tag), name_(Http::LowerCaseString(request_header.name())),
+      default_value_(request_header.default_value()) {}
+
+absl::string_view RequestHeaderCustomTag::value(const CustomTagContext& ctx) const {
+  if (!ctx.request_headers) {
+    return default_value_;
+  }
+  const Http::HeaderEntry* entry = ctx.request_headers->get(name_);
+  return entry ? entry->value().getStringView() : default_value_;
+}
+
+MetadataCustomTag::MetadataCustomTag(const std::string& tag,
+                                     const envoy::type::tracing::v2::CustomTag::Metadata& metadata)
+    : GeneralCustomTag(tag), metadata_key_(metadata.metadata_key()),
+      default_value_(metadata.default_value()) {}
+
+void MetadataCustomTag::apply(Span& span, const CustomTagContext& ctx) const {
+  const envoy::api::v2::core::Metadata* meta = metadata(ctx);
+  if (!meta) {
+    if (!default_value_.empty()) {
+      span.setTag(tag(), default_value_);
+    }
+    return;
+  }
+  const ProtobufWkt::Value& value = Envoy::Config::Metadata::metadataValue(*meta, metadata_key_);
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::kBoolValue:
+    span.setTag(tag(), value.bool_value() ? "true" : "false");
```

Comment:
General note - we might want to consider supporting other primitive types for setting tags in the future.

Reply:
Like `NullValue`?

Reply:
Yes - but not in this PR, just was mentioning as a possible future improvement.